### PR TITLE
(QA-3415) - First pass at adding perf assertions

### PR DESCRIPTION
**Pull request by pinkypie**
Created GatlingResult object to make things tidier in the tests.
Added push to BigQuery functionality and baseline comparison functionality.

* Will also require a change to beaker-benchmark to return the @@session_timestamp from the start_monitoring method - https://github.com/puppetlabs/beaker-benchmark/pull/7

### CODE REVIEW

File: tests/ApplesToApples.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+
+  gatlingassertions = "SUCCESSFUL_REQUESTS=100 " + "MAX_RESPONSE_TIME_AGENT=20000 "  + "TOTAL_REQUEST_COUNT=70 "
+
+  # pass in gatling scenario file name and simulation id
+  perf_setup('ApplesToApples.json','PerfTestLarge', gatlingassertions)
+
+  result = perf_result
+
+  step 'max response' do
```
Comment (samwoods1) [inline tests/ApplesToApples.rb]: I think we also decided to add an assertion for average response?

File: simulation-runner/src/main/scala/com/puppetlabs/gatling/runner/ConfigDrivenSimulation.scala [hunk 2]
```diff
@@ -109,6 +109,14 @@ class ConfigDrivenSimulation extends Simulation {
       .protocols(httpProtocol)
   })
 
-  setUp(scns)
 //  scns.foreach(setUp(_))
+    def getEnvVar(varName: String): String = {
```
[the remark this answers is not in the document]
Comment (pinkypie) [reply simulation-runner/src/main/scala/com/puppetlabs/gatling/runner/ConfigDrivenSimulation.scala]: Working on this.

File: tests/ApplesToApples.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      assert(false, "avg_mem was not within 10% of baseline.")
+    end
+
+    if (atop_result.avg_disk_write.to_f - baseline_dsk_write) / baseline_dsk_write * 100 <= 10 then
+      assert(true, "avg_dsk_write passed")
+    else
+      assert(false, "avg_dsk_write was not within 10% of baseline.")
+    end
+
+    if (gatling_result.avg_response_time.to_f - baseline_avg_resp_time) / baseline_avg_resp_time * 100 <= 10 then
```
Comment (samwoods1) [inline tests/ApplesToApples.rb]: same as above

File: tests/acceptance.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+  #This step will only be run if BASELINE_PE_VER and has been set.
+  step 'baseline assertions' do
+
+    baseline_cpu = baseline_result.baseline_cpu.to_f
+    baseline_memory = baseline_result.baseline_memory.to_f
+    baseline_dsk_write = baseline_result.baseline_memory.to_f
+    baseline_avg_resp_time = baseline_result.baseline_avg_resp_time.to_f
+
+    if (atop_result.avg_cpu.to_f - baseline_cpu) / baseline_cpu * 100 <= 10 then
```
Comment (samwoods1) [inline tests/acceptance.rb]: Same comments as from apples to apples on assertions below

File: tests/helpers/perf_run_helper.rb [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    eval(ENV['PUSH_TO_BIGQUERY'])
+  end
+
+  def push_to_bigquery
+    bigquery = Google::Cloud::Bigquery.new project: "perf-metrics"
+    dataset = bigquery.dataset "perf_metrics"
+    atop_table = dataset.table "atop_metrics"
+
+    row = [{
+               #"insertId" => Random.new.rand(1..100),
```
Comment (samwoods1) [inline tests/helpers/perf_run_helper.rb]: Can this be removed?

File: NEW_README.md [hunk 6]
```diff
@@ -23,6 +23,35 @@ You can execute the 'acceptance' rake task which will run everything in VMPooler
 
 `ABS_OS=centos-7-x86-64-west BEAKER_HOSTS=config/beaker_hosts/pe-perf-test.cfg`
 
+### Assertions and Baseline Comparisons
+
+In order to use the BigQuery integration, you will need to perform the following steps:
+
+* Choose to create a new key for the perf-metrics service account - https://console.developers.google.com/iam-admin/serviceaccounts/project?project=perf-metrics&organizationId=635869474587
```
[the remark this answers is not in the document]
Comment (samwoods1) [reply NEW_README.md]: Still get the same error.

File: NEW_README.md [hunk 7]
```diff
@@ -23,6 +23,35 @@ You can execute the 'acceptance' rake task which will run everything in VMPooler
...
+`export BASELINE_PE_VER=`
+
+to the version of PE you want to compare against.
+
+Things to note:
+
+* If any of the test specific assertions fail, the baseline comparison assertions will not be executed.
+* If BASELINE_PE_VER has not been set than the baseline comparison assertions in the last test step will not be run (test step will be skipped).
+* Whatever string you pass to the job to indicate the BEAKER_PE_VER needs to be used consistently. For example, if you set the version in the format `2018.1.1-rc0-11-g8fbde83` you need to use this again as the BASELINE_PE_VER or the test will error/not return expected results.
+* If BASELINE_PE_VER is not found in BigQuery then the last step will error.
```
[the remark this answers is not in the document]
Comment (pinkypie) [reply NEW_README.md]: You should be able to access (I gave the whole Silver team admin rights) - https://bigquery.cloud.google.com/table/perf-metrics:perf_metrics.atop_metrics
And run:
SELECT pe_build_number FROM [perf-metrics:perf_metrics.atop_metrics]
GROUP BY pe_build_number

File: tests/ApplesToApples.rb [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  #This step will only be run if BASELINE_PE_VER and has been set.
+  step 'baseline assertions' do
+
+    baseline_cpu = baseline_result.baseline_cpu.to_f
+    baseline_memory = baseline_result.baseline_memory.to_f
+    baseline_dsk_write = baseline_result.baseline_memory.to_f
+    baseline_avg_resp_time = baseline_result.baseline_avg_resp_time.to_f
+
+    if (atop_result.avg_cpu.to_f - baseline_cpu) / baseline_cpu * 100 <= 10 then
+      assert(true, "avg_cpu passed")
```
[the remark this answers is not in the document]
Comment (pinkypie) [reply tests/ApplesToApples.rb]: Yes of course! 🤣

**Comment by samwoods1** [comment]
Thanks @pinkypie ! I was able to run with and without the baseline comparison, and store a new baseline.

**Pull request merged**